core/src/main/java/io/questdb/cairo/vm/api/MemoryCARW.java

```diff
@@ -0,0 +1,247 @@
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.cairo.vm.api;
+
+import io.questdb.cairo.CairoException;
+import io.questdb.cairo.TableUtils;
+import io.questdb.cairo.vm.Vm;
+import io.questdb.std.*;
+
+public interface MemoryCARW extends MemoryCR, MemoryARW, MemoryCA, MemoryMAT {
+
+    default long putBin(BinarySequence value) {
+        if (value != null) {
+            final long len = value.length();
+            long addr = appendAddressFor(len + Long.BYTES);
+            Unsafe.getUnsafe().putLong(addr, len);
+            value.copyTo(addr + Long.BYTES, 0, len);
+            return getAppendOffset();
+        }
+        return putNullBin();
+    }
+
+    default long putBin(long from, long len) {
+        if (len > 0) {
+            long addr = appendAddressFor(len + Long.BYTES);
+            Unsafe.getUnsafe().putLong(addr, len);
+            Vect.memcpy(addr + Long.BYTES, from, len);
+            return getAppendOffset();
+        }
+        return putNullBin();
+    }
+
+    @Override
+    default long getAddress() {
+        return getPageAddress(0);
+    }
+
+    @Override
+    default void putBlockOfBytes(long from, long len) {
+        Vect.memcpy(appendAddressFor(len), from, len);
+    }
+
+    default void putBool(boolean value) {
+        putByte((byte) (value ? 1 : 0));
+    }
+
+    default void putByte(byte b) {
+        Unsafe.getUnsafe().putByte(appendAddressFor(Byte.BYTES), b);
+    }
+
+    default void putChar(char value) {
+        Unsafe.getUnsafe().putChar(appendAddressFor(Character.BYTES), value);
+    }
+
+    default void putDouble(double value) {
+        Unsafe.getUnsafe().putDouble(appendAddressFor(Double.BYTES), value);
+    }
+
+    default void putFloat(float value) {
+        Unsafe.getUnsafe().putFloat(appendAddressFor(Float.BYTES), value);
+    }
+
+    default void putInt(int value) {
+        Unsafe.getUnsafe().putInt(appendAddressFor(Integer.BYTES), value);
+    }
+
+    default void putLong(long value) {
+        Unsafe.getUnsafe().putLong(appendAddressFor(Long.BYTES), value);
+    }
+
+    default void putLong128(long l1, long l2) {
+        long addr = appendAddressFor(16);
+        Unsafe.getUnsafe().putLong(addr, l1);
+        Unsafe.getUnsafe().putLong(addr + 8, l2);
+    }
+
+    default void putLong256(long l0, long l1, long l2, long l3) {
+        final long addr = appendAddressFor(32);
+        Unsafe.getUnsafe().putLong(addr, l0);
+        Unsafe.getUnsafe().putLong(addr + Long.BYTES, l1);
+        Unsafe.getUnsafe().putLong(addr + Long.BYTES * 2, l2);
+        Unsafe.getUnsafe().putLong(addr + Long.BYTES * 3, l3);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).